### ISSUE TITLE
Trouble serving lambdas and static site on Windows (bash and cmd)

### ISSUE THREAD

**Issue by jugglingcats** [awslabs/aws-sam-local#145]
I've having trouble getting both lambda function via gateway and a static site served together.

*cmd shell*
By default `sam local start-api` without any parameters gives an `invalid bind mount spec` error, because the mount spec is in Windows form: `C:\Users\blah`.

If I specify `--docker-volume-basedir /c/Users/blah` then the lambda works fine, but now the command output contains `Mounting static files from \c\Users\blah\public at /`, and this static dir is not valid on a Windows machine (404 errors hitting urls).

If I try to specify `--static-dir` using the Windows full path (or any path) it simply concatenates this with the `--docker-volume-basedir` and again we get an invalid path, see:

https://github.com/awslabs/aws-sam-local/blob/b3101e5ea50467305d1478d1cff6709e361256fc/start.go#L53-L56

https://github.com/awslabs/aws-sam-local/blob/b3101e5ea50467305d1478d1cff6709e361256fc/start.go#L120-L125

*Git bash shell*
Same behaviour as cmd shell without any parameters. If I try to specify `--docker-volume-basedir /c/Users/blah` it doesn't really do anything: this path is converted to Windows format quite early on (not sure by what?) and again we get `invalid bind mount spec`.

I've tried playing with `COMPOSE_CONVERT_WINDOWS_PATHS` but it doesn't seem to make any difference in either shell.

It seems to me that `--docker-volume-basedir` and `--static-dir` need to be made more independent, since one is consumed by docker and other by the "host" HTTP server.

Any help much appreciated!

**Comment by alexb-uk** [comment]
Note: I only started playing with SAM Local a couple of days ago and don't yet have a fully working setup BUT I believe I've managed to get slightly further :)

I'm running 
- Windows 10 Pro
- Docker version 17.06.2-ce, build cec0b72
- sam version 0.2.2

For me the API server starts fine and serves events but they fail with a strange parse error:
````bash
alexb@DESKTOP-5HFQQQL MINGW64 ~/PhpstormProjects/swaggerizeTest (master)
$ sam local start-api
2017/10/02 09:43:13 Connected to Docker 1.30
2017/10/02 09:43:13 Fetching lambci/lambda:nodejs6.10 image for nodejs6.10 runtime...
nodejs6.10: Pulling from lambci/lambda
Digest: sha256:7eb4ced6a15ae3c30effc4ec0cd3aabb2bd57c9a8330b37920c3d5d722d81083
Status: Image is up to date for lambci/lambda:nodejs6.10

Mounting index.handler (nodejs6.10) at http://127.0.0.1:3000/Property-API/0.0.1/api-docs [GET]
Mounting static files from public at /

You can now browse to the above endpoints to invoke your functions.
You do not need to restart/reload SAM CLI while working on your functions,
changes will be reflected instantly/automatically. You only need to restart
SAM CLI if you update your AWS SAM template.

2017/10/02 09:43:20 Invoking index.handler (nodejs6.10)
START RequestId: 1241967b-5342-17a2-79d5-7d76b518bc81 Version: $LATEST
2017-10-02T08:43:22.361Z        1241967b-5342-17a2-79d5-7d76b518bc81    Registering plugins!!!!
2017-10-02T08:43:22.712Z        1241967b-5342-17a2-79d5-7d76b518bc81    Setting Host to 4b2cfc67527a:8000
END RequestId: 1241967b-5342-17a2-79d5-7d76b518bc81
REPORT RequestId: 1241967b-5342-17a2-79d5-7d76b518bc81  Duration: 1241.16 ms    Billed Duration: 0 ms   Memory Size: 0 MB       Max Memory Used: 59 MB
2017/10/02 09:43:22 Function returned an invalid response (must include one of: body, headers or statusCode in the response object): invalid character '-' after top-level value
````

The contents of my template.yml is quite basic:
````bash
AWSTemplateFormatVersion : '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Resources:
  Hapihello:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.handler
      Runtime: nodejs6.10
      Events:
        ApiDocs:
          Type: Api
          Properties:
            Path: /Property-API/0.0.1/api-docs
            Method: get
````

How does that compare to your template??

Is there anything you can think of to compare our two environments?

Cheers

Alex

**Comment by jugglingcats** [comment]
Hi @alexb-uk I'd guess your lambda function is not responding with API gateway style. You need to return something like this:

```
 exports.get = function(event, context) {
   context.succeed({
     statusCode: 200,
     body: "Hello World",
     headers: {'Content-Type': 'text/html'}
   });
 };
```

**Comment by alexb-uk** [comment]
_**Note**: I didn't want to hijack your post so have created my own issue regarding the weird parsing problem. But thanks very much for the suggestion._

I did double check that and even tried hardcoding a response rather than using the Hapi server but didn't seem to make a difference. I might try again now that I've slept between last attempt.

However my Lambda function IS working perfectly on the real AWS Lambda service. So I assumed this with a SAM Local config issue rather than Javascript code issue.

**Comment by jugglingcats** [comment]
Ok, thanks for creating own issue as I don't think your issue and mine are the same.

My issue is not ability to serve lambda function (which is working fine for me), it's the ability to do this _and_ serve static files from docroot. I can do one or other but not both at the same time :(

Good luck with your issue!

**Comment by Liamdelee** [comment]
just a question so i get it right.
so you cannot run `sam local start-api` without an error? because i myself run sam local on windows 10 and have no problem with it. At the moment i have a MongoDB local running, use AWS sam local for Lambda and API gateway, and run a webpage (just an index that will show my books in DynamoDB and that can add books to DynamoDB).

**Comment by jugglingcats** [comment]
Hi @Liamdelee, I can run `sam local start-api` no problem. I can get output from my lambda (which returns HTML content), and I can separately get static files served (from public subdir of my project), but I cannot get them both together!

Is your project serving lambda output and static files (like CSS, JS, etc)? My project is the starter lambda nodejs project from AWS CodeStar.

Are you using options `--docker-volume-basedir` or `--static-dir` as arguments?

Thanks.

**Comment by Liamdelee** [comment]
oohh okay, i am running everything as seperate projects, so my sam local is a folder and my static webpage is a folder that i run on two diffrent instances. so you want to run sam local and your static webpage with the same call or? or can you just not run sam local and your static webpage at the same time?

**Comment by jugglingcats** [comment]
@Liamdelee maybe if I explain how I got here it would help.

My starting point is the CodeStar template for nodejs lambda application. This is successfully deployed here: https://ze4phbp3xh.execute-api.eu-west-2.amazonaws.com/Prod.

This consists of lambda code and static content on S3, and source code in AWS code commit. Works great but takes a while to deploy changes, so my goal is to work with sam local on this project, ideally with the same structure.

The project structure looks like this:
![image](https://user-images.githubusercontent.com/526349/31084258-dce422dc-a78b-11e7-8acd-3dc265a82f70.png)

You can see in the screen grab that the lambda reads index.html and serves it. All the asset references in the HTML page are relative, eg: `<link href="assets/css/styles.css" rel="stylesheet">`. As an aside, during deployment, it's the buildspec.yml that re-writes these relative urls to include the full URL where the S3 bucket is served: https://s3-eu-west-2.amazonaws.com/aws-codestar-eu-west-2-986339458627-club-app/public/assets in my case. I thought this was some lambda/codestar magic until I found the `sed` command :)

I order to run locally without editing the HTML, I need the following to work:

GET http://localhost:3000 (should serve the lambda, resulting in HTML)
GET http://localhost:3000/assets (should serve static content from `public` subdir)

It seems that aws local should support this with the various options but it's giving me problems on Windows. I plan to try the same setup on Linux to see if it works, but am pretty sure it should!

Of course I could use a local proxy, or some other hack around and would be happy to do this, but the aws local folks clearly intend for this to work ootb.

**Comment by jugglingcats** [comment]
Re-created my scenario in an Ubuntu VM and it works as expected... At least the lambda and static assets are served correctly. It takes 20s to serve the lambda but that must be a different issue!

**Comment by Liamdelee** [comment]
okay i get where you are going now srry for the misunderstanding from my part.
So if i get it right he is pulling the html page from the S3 bucket and that is working great but he won't find the other files in let's say the assets/css folder?
do you get some kind of error in logs or did you try consol.log() to see where the codes goes and stops? sounds stupid and takes some time but helped me a lot already :).
i never did this to pull from S3 and serve the content so i would try and find out if it can 
A: reatch the code.
B: where the execution fails. 
and hope somebody here can help you out :)
hope you can make it work and keep me update because i was also on planning to create something like this :)

**Comment by sanathkr** [comment]
@jugglingcats Sorry for the whole trouble in getting this working in Windows. One of us will dive deep.

Regarding the 20s issue, see https://github.com/awslabs/aws-sam-local/pull/159. Can you configure AWS credentials and check? It should fix your problem

**Comment by jugglingcats** [comment]
Hi @sanathkr, yes I sorted the 20s problem using credentials. All working fine on Linux.

I've actually been using create-react-app in combination with sam local on Windows and it runs webpack dev server, which has a simple proxy option for API requests, and this is working well. So getting sam local to serve static assets is less of an issue for me now...

I'm happy to test any fixes, but have been unable to build sam local on Windows due to long filename woes. Windoze!

**Comment by PaulMaddox** [comment]
This fix is included in the latest release (v0.2.3).
Use `npm update -g aws-sam-local` to upgrade.
